### PR TITLE
refactor TRD digit class

### PR DESCRIPTION
**Pull request by bazinski**
- Convert digits to store rob/mcm and not pad/row.
- The interface for pad row remains so code wont change.
- For shared pads the lower indexed mcm is returned. It is the users responsibility to check if its shared if needed for their purposes.
 - The plan was to move digit to DataFormats, but there is a circular library, so postponed to later.

### CODE REVIEW

File: Detectors/TRD/base/include/TRDBase/Digit.h [hunk 1]
```diff
@@ -30,36 +30,81 @@ namespace trd
...
-  int getPad() const { return mPad; }
+  int getDetector() const { return mRob; }
+  int getRow() const { return FeeParam::getPadRowFromMCM(mRob, mMCM); }
+  int getPad() const { return FeeParam::getPadColFromADC(mRob, mMCM, mChannel); }
+  int getRob() const { return mRob; }
+  int getMCM() const { return mMCM; }
+  int getChannel() const { return mChannel; }
+  int isSharedDigit()
+  {
+    if (mChannel == 0 || mChannel == 18 || mChannel == 19)
```
Comment (tdietel) [inline Detectors/TRD/base/include/TRDBase/Digit.h]: I think this should be 0, 19 or 20.

File: Detectors/TRD/base/src/FeeParam.cxx [hunk 2]
```diff
@@ -195,7 +195,7 @@ void FeeParam::Copy(FeeParam& p) const
 }
 
 //_____________________________________________________________________________
-int FeeParam::getPadRowFromMCM(int irob, int imcm) const
+int FeeParam::getPadRowFromMCM(int irob, int imcm)
```
Comment (tdietel) [inline Detectors/TRD/base/src/FeeParam.cxx:198]: Minor style question: why did you remove the const?

File: Detectors/TRD/simulation/src/Digitizer.cxx [hunk 3]
```diff
@@ -519,7 +519,7 @@ bool Digitizer::convertSignalsToADC(SignalContainer& signalMapCont, DigitContain
       adcs[tb] = adc;
     } // loop over timebins
     // Convert the map to digits here, and push them to the container
-    digits.emplace_back(det, row, col, adcs, getEventTime());
+    digits.emplace_back(det, row, col, adcs);
```
Comment (tdietel) [inline Detectors/TRD/simulation/src/Digitizer.cxx:522]: The removal of the event time might be an issue for @jolopezl - please confirm.

**Review by tdietel — CHANGES_REQUESTED** [review]
I think the indicator for the shared pads is buggy and should be fixed.

File: Detectors/TRD/base/include/TRDBase/Digit.h [hunk 4]
```diff
@@ -18,8 +18,8 @@
 #include <numeric>
 #include "Rtypes.h" // for ClassDef
 
-#include "CommonDataFormat/TimeStamp.h"
-#include "TRDBase/CommonParam.h"
+//#include "TRDBase/CommonParam.h"
```
Comment (jolopezl) [inline Detectors/TRD/base/include/TRDBase/Digit.h]: I prefer to remove the additional header instead of keeping it commented.

File: Detectors/TRD/base/include/TRDBase/Digit.h [hunk 5]
```diff
@@ -30,36 +30,81 @@ namespace trd
...
+    setMCM(mcm);
+    setChannel(channel);
+    setADC(adc);
+  }
+  Digit(const int det, const int rob, const int mcm, const int channel) // add adc data in a seperate step
+  {
+    setRob(rob);
+    setMCM(mcm);
+    setChannel(channel);
+  }
```
Comment (jolopezl) [inline Detectors/TRD/base/include/TRDBase/Digit.h]: In the past, we have been obligated to separate the implementation from the header. Since we already have a source file for the Digit class, I suggest we put this in `src/Digit.cxx`, but IMO, it's more than fine to keep it here so that you can consider keeping this right here.

File: Detectors/TRD/simulation/src/Digitizer.cxx (hunk 3 above)
Comment (jolopezl) [reply Detectors/TRD/simulation/src/Digitizer.cxx:522]: Looks good to me.

File: Detectors/TRD/base/include/TRDBase/Digit.h (hunk 1 above)
Comment (bazinski) [reply Detectors/TRD/base/include/TRDBase/Digit.h]: Wow where did i get that from..... fixed.

File: Detectors/TRD/simulation/src/Digitizer.cxx (hunk 3 above)
Comment (bazinski) [reply Detectors/TRD/simulation/src/Digitizer.cxx:522]: This came out when we originally discussed this. The above change is actually a patch from @jolopezl but he can confirm

File: Detectors/TRD/base/src/FeeParam.cxx (hunk 2 above)
Comment (bazinski) [reply Detectors/TRD/base/src/FeeParam.cxx:198]: I seem to remember it was a hack because I got frustrated with something in the digit class. Maybe let me go back and fix that properly.

**Review by jolopezl — COMMENTED** [review]
I am happy with the Digit class changes.

Just a minor style comment: I prefer to read ROB instead of Rob, as in the Tracklet class (`mROB`, `getROB`, `setROB`), to make the code more consistent and easier to browse.

File: Detectors/TRD/simulation/src/Digitizer.cxx (hunk 3 above)
Comment (jolopezl) [reply Detectors/TRD/simulation/src/Digitizer.cxx:522]: I confirm.

File: Detectors/TRD/base/include/TRDBase/Digit.h (hunk 5 above)
Comment (bazinski) [reply Detectors/TRD/base/include/TRDBase/Digit.h]: Done

File: Detectors/TRD/base/include/TRDBase/Digit.h [hunk 6]
```diff
@@ -30,36 +29,55 @@ namespace trd
...
-  std::uint8_t mRow{0};       // pad row, 0-15
-  std::uint8_t mPad{0};       // pad within pad row, 0-143
-  ArrayADC mADC{};            // ADC vector (30 time-bins)
+  std::uint16_t mDetector{0}; // detector, the chamber
+  std::uint16_t mROB{0};      // read out board with in chamber
+  std::uint8_t mMCM{0};       // MCM chip this digit is attached to
+  std::uint8_t mChannel{0};   // channel of this chip the digit is attached to.
+
+  ArrayADC mADC{}; // ADC vector (30 time-bins)
   ClassDefNV(Digit, 2);
```
Comment (shahor02) [inline Detectors/TRD/base/include/TRDBase/Digit.h]: could you increment the class version?

File: Detectors/TRD/base/src/FeeParam.cxx [hunk 7]
```diff
@@ -557,7 +557,7 @@ int FeeParam::getORIfromHCID(int hcid) const
   return getORIinSM(detector, side); // it takes readoutboard but only cares if its odd or even hence side here.
   return 1;
```
Comment (shahor02) [inline Detectors/TRD/base/src/FeeParam.cxx]: 2nd return is hardly needed?

**Review by shahor02 — CHANGES_REQUESTED** [review]

**Comment by bazinski** [comment]
@shahor02 shot for seeing, now fixed.

File: Detectors/TRD/base/include/TRDBase/Digit.h [hunk 8]
```diff
@@ -30,37 +29,56 @@ namespace trd
...
   // Copy
   Digit(const Digit&) = default;
   // Assignment
   Digit& operator=(const Digit&) = default;
   // Modifiers
+  void setROB(int rob) { mROB = rob; }
+  void setROB(int row, int pad) { mROB = FeeParam::getROBfromPad(row, pad); }
+  void setMCM(int mcm) { mMCM = mcm; }
+  void setMCM(int row, int pad) { mMCM = FeeParam::getMCMfromPad(row, pad); }
+  void setChannel(int channel) { mChannel = channel; }
```
Comment (tdietel) [inline Detectors/TRD/base/include/TRDBase/Digit.h:59]: Are these setters actually used? I would expect that we always know at construction which pad these digits refer to.

File: Detectors/TRD/base/include/TRDBase/Digit.h [hunk 9]
```diff
@@ -30,37 +29,56 @@ namespace trd
...
   ADC_t getADCsum() const { return std::accumulate(mADC.begin(), mADC.end(), (ADC_t)0); }
 
  private:
-  std::uint16_t mDetector{0}; // TRD detector number, 0-539
-  std::uint8_t mRow{0};       // pad row, 0-15
-  std::uint8_t mPad{0};       // pad within pad row, 0-143
-  ArrayADC mADC{};            // ADC vector (30 time-bins)
-  ClassDefNV(Digit, 2);
+  std::uint16_t mDetector{0}; // detector, the chamber
+  std::uint16_t mROB{0};      // read out board with in chamber
```
Comment (tdietel) [inline Detectors/TRD/base/include/TRDBase/Digit.h]: In principle, this could be uint8_t

**Review by tdietel — DISMISSED** [review]
I added one or two more comments, but I think we can commit as it is.

**Comment by bazinski** [comment]
@tdietel let me change the rob value to uint8_t that will have somewhat of an impact on the sizing of the output, assuming it does not get padded.

**Comment by bazinski** [comment]
@tdietel regarding the set methods, i dont use them, they are simply there for completeness.

**Comment by shahor02** [comment]
> @tdietel let me change the rob value to uint8_t that will have somewhat of an impact on the sizing of the output, assuming it does not get padded.

It will be padded

File: Detectors/TRD/base/include/TRDBase/Digit.h (hunk 8 above)
Comment (jolopezl) [reply Detectors/TRD/base/include/TRDBase/Digit.h:59]: I prefer keeping them for completeness; they are harmless.

**Review by jolopezl — APPROVED** [review]

**Comment by bazinski** [comment]
@shahor02 when you get a chance could you approve the change as per your request and merge?

**Review by shahor02 — APPROVED** [review]
